### PR TITLE
Feat(web-react): Introduce ScrollView component #DS-464

### PR DESCRIPTION
**Pull request by tomassychra**
# Introduce ScrollView component [DS-464](https://jira.lmc.cz/browse/DS-464)

### CODE REVIEW

File: packages/web-react/src/components/ScrollView/useScrollPosition.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      bottom: contentPosition.bottom - viewportPosition.bottom,
+      left: contentPosition.left - viewportPosition.left,
+      right: contentPosition.right - viewportPosition.right,
+      top: contentPosition.top - viewportPosition.top,
+    };
+  };
+
+  const handleScrollViewState = () => {
+    const isDirectionHorizontal = direction === 'horizontal';
+    const scrollPositionStart = isDirectionHorizontal ? 'left' : 'top';
```
Comment (literat) [inline packages/web-react/src/components/ScrollView/useScrollPosition.ts:51]: I think this is the same as I wrote a comment here: https://github.com/lmc-eu/spirit-design-system/pull/844/files#r1203601435

It would be great if we can use dictionaries here, like `Position.LEFT` instead of just `left`. We can then use this dictionary to define the keys of the current position object and so on.

So there should be two dictionaries, one with directions and one with alignments.

File: packages/web-react/src/types/scrollView.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { ChildrenProps, StyleProps } from './shared';
+
+export type ScrollViewDirectionType = 'horizontal' | 'vertical';
```
Comment (literat) [inline packages/web-react/src/types/scrollView.ts:3]:
```suggestion
export type ScrollViewDirectionType = Direction.HORIZONTAL | Direction.VERTICAL;
```

File: packages/web/src/js/ScrollView.ts
Comment (literat) [inline packages/web/src/js/ScrollView.ts:1]: @tomassychra I think you have wrongly set the target branch because I was expecting, that the web package code will not be part of this PR.

**Review by literat — CHANGES_REQUESTED** [review]
I have only a few suggestions about type and constants definitions that should be the same in both packages. Also I see the changes from web package, please check the target branch.

**Comment by literat** [comment]
https://deploy-preview-849--spirit-design-system-storybook.netlify.app/

File: packages/web-react/src/components/ScrollView/index.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export { default as ScrollView } from './ScrollView';
+export * from './useScrollViewStyleProps';
+export * from './useScrollPosition';
+export * from './const';
```
[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/index.ts:4]: fixed

File: packages/web-react/src/components/ScrollView/useScrollPosition.ts (hunk 1 above)
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/useScrollPosition.ts:51]: done

File: packages/web-react/src/types/scrollView.ts (hunk 2 above)
Comment (tomassychra) [reply packages/web-react/src/types/scrollView.ts:3]: done

File: packages/web/src/js/ScrollView.ts
Comment (tomassychra) [reply packages/web/src/js/ScrollView.ts:1]: How do you mean wrong? I created a branch from main and then rebased `feature/DS-449-web-scroll-view`

File: packages/web-react/src/constants/dictionaries.ts [hunk 4]
```diff
@@ -1,4 +1,11 @@
 /* Alignment */
+export const Alignment = {
```
Comment (crishpeen) [inline packages/web-react/src/constants/dictionaries.ts:2]: This is not in our dictionary. 

Either remove it from here and make it local for the component or add it to the notion and global dictionaries docs in the repo root, but we should be all aware of that (including Michal).

File: packages/web-react/src/types/shared/dictionaries.ts [hunk 5]
```diff
@@ -28,6 +33,10 @@ export type EmotionColorsDictionaryType<C = undefined> = (typeof EmotionColors)[
 export type TextColorsDictionaryKeys = keyof typeof TextColors;
 export type TextColorsDictionaryType<C = undefined> = (typeof TextColors)[TextColorsDictionaryKeys] | C;
 
+/* Direction */
+export type DirectionDictionaryKeys = keyof typeof Direction;
```
Comment (crishpeen) [inline packages/web-react/src/types/shared/dictionaries.ts:37]: This also isn't in our shared dictionaries. We might add that, but it should be discussed with Michal too, because he has to be aware of that and name the props in Figma like that.

File: packages/web-react/src/components/ScrollView/README.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# ScrollView
+
+## Usage
+
+### Default (Horizontal)
```
Comment (crishpeen) [inline packages/web-react/src/components/ScrollView/README.md:5]: That's not true. Use set vertical as default. 
```suggestion
### Default (Vertical)
```

File: packages/web-react/src/components/ScrollView/README.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+We need to set the maximum height if it is not defined by the parent.
+
+```javascript
+<div style={{ height: '250px' }}>
```
[the remark this answers is not in the document]
Comment (crishpeen) [reply packages/web-react/src/components/ScrollView/README.md:10]: I think it is mainly confusing because there is a wrong title here. Also, we don't need a wrapper. Because there are several ways to limit the height, it could be fixed height like this or it could be in some kind of grid layout etc. This should be explained that we set inline height only to showcase the ScrollView.

File: packages/web-react/src/components/ScrollView/README.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede
+      mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus.
+      Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis,
+      feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet.
+    </p>
+  </ScrollView>
+</div>
+```
+
+### Vertical
```
Comment (crishpeen) [inline packages/web-react/src/components/ScrollView/README.md:25]:
```suggestion
### Horizontal
```

File: packages/web-react/src/components/ScrollView/README.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+```javascript
+<ScrollView>
```
Comment (crishpeen) [inline packages/web-react/src/components/ScrollView/README.md:28]:
```suggestion
<ScrollView direction="horizontal">
```

File: packages/web-react/src/components/ScrollView/README.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  </p>
+</ScrollView>
+```
+
+## Accordion Props
+
+| Prop name          | Type                       | Default      | Required | Description                 |
+| ------------------ | -------------------------- | ------------ | -------- | --------------------------- |
+| `children`         | `ReactNode`                | -            | yes      | ScrollView children's nodes |
+| `direction`        | `'horizontal', 'vertical'` | `'vertical'` | no       | Direction of the wrapper    |
```
Comment (crishpeen) [inline packages/web-react/src/components/ScrollView/README.md:46]:
```suggestion
| `direction`        | `horizontal`, `vertical` | `vertical` | no       | Direction of the scroll |
```

File: packages/web-react/src/components/ScrollView/README.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec,
+    vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede
+    mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus.
+    Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis,
+    feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet.
+  </p>
+</ScrollView>
+```
+
+## Accordion Props
```
Comment (crishpeen) [inline packages/web-react/src/components/ScrollView/README.md:41]: ? :)

File: packages/web-react/src/components/ScrollView/README.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+| Prop name          | Type                       | Default      | Required | Description                 |
```
Comment (crishpeen) [inline packages/web-react/src/components/ScrollView/README.md:43]: You are missing the isScrollbarDisabled prop. Also in the implementation.

**Review by crishpeen — CHANGES_REQUESTED** [review]

File: packages/web/src/js/ScrollView.ts
Comment (crishpeen) [reply packages/web/src/js/ScrollView.ts:1]: I didn't realize that yesterday when we talked about this. But now I realized it is hard to do CR as there is also Adam's commit and it is in general confusing. I didn't have arguments yesterday, but now it is obvious. See how I do it in the Twig ScrollView. I place my branch from Adam's branch so in my PR there are only my commits. 

Yes, we can filter in the files to only your commit, but we can do it only to one specific (if there were more of them, we would have to do it separately) and then the comment options are limited.

I suggest you rebase your branch over current Adam's and direct this PR into his branch so there are only your commits.

Comment (literat) [reply packages/web/src/js/ScrollView.ts:1]: @tomassychra The base branch in this PR is set to the`main`. Right top corner of the page -> `Edit` -> Change branch to Adam's.

File: packages/web-react/src/components/ScrollView/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (crishpeen) [reply packages/web-react/src/components/ScrollView/README.md:10]: With small help from our AI colleagues
> It's crucial to understand that the ScrollView component, like any HTML element, expands to fit its content. Without a height limit on the parent container, the ScrollView will stretch to include all its children, which could negate the need for scrolling.

> To make scrolling and scroll indicators function correctly, the parent container's height must be limited. In our examples, we set this height limit using inline styles for demonstration purposes only.

[the remark this answers is not in the document]
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/README.md:10]: done in README

File: packages/web/src/js/ScrollView.ts
Comment (tomassychra) [reply packages/web/src/js/ScrollView.ts:1]: outdated now

File: packages/web-react/src/components/ScrollView/README.md (hunk 6 above)
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/README.md:5]: fixed

File: packages/web-react/src/components/ScrollView/README.md (hunk 8 above)
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/README.md:25]: fixed

File: packages/web-react/src/components/ScrollView/README.md (hunk 9 above)
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/README.md:28]: fixed

File: packages/web-react/src/components/ScrollView/README.md (hunk 12 above)
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/README.md]: done

File: packages/web-react/src/types/shared/dictionaries.ts (hunk 5 above)
Comment (tomassychra) [reply packages/web-react/src/types/shared/dictionaries.ts]: fixed

File: packages/web-react/src/constants/dictionaries.ts (hunk 4 above)
Comment (tomassychra) [reply packages/web-react/src/constants/dictionaries.ts]: fixed

File: packages/web-react/src/types/shared/dictionaries.ts [hunk 13]
```diff
@@ -9,7 +9,6 @@ import {
   ValidationStates,
 } from '../../constants';
 
-/* Alignment */
```
Comment (crishpeen) [inline packages/web-react/src/types/shared/dictionaries.ts:12]: We probably want to keep that (but it should actually be `Alignment X`)

File: packages/web-react/src/types/scrollView.ts (hunk 2 above)
Comment (crishpeen) [reply packages/web-react/src/types/scrollView.ts:3]: I still see the previous values?

File: packages/web-react/src/components/ScrollView/__tests__/useScrollViewStyleProps.test.ts [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+      }),
+    );
+
+    expect(result.current.classProps).toBeDefined();
+    expect(result.current.classProps.root).toBe(
+      'ScrollView ScrollView--horizontal ScrollView--scrollbarDisabled is-scrolled-at-end is-scrolled-at-start',
+    );
+    expect(result.current.classProps.viewport).toBe('ScrollView__viewport');
+    expect(result.current.classProps.content).toBe('ScrollView__content');
+    expect(result.current.classProps.indicators).toBe('ScrollView__indicators ScrollView__indicators--shadows');
```
Comment (crishpeen) [inline packages/web-react/src/components/ScrollView/__tests__/useScrollViewStyleProps.test.ts:22]: I would love to see test to test `both` too, but 🤷

**Review by crishpeen — APPROVED** [review]

File: packages/web-react/src/types/shared/dictionaries.ts (hunk 13 above)
Comment (tomassychra) [reply packages/web-react/src/types/shared/dictionaries.ts:12]: It could be also AlignmentY. Below are similar colors with a uniform caption.

File: packages/web-react/src/components/ScrollView/ScrollView.tsx [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+import { useStyleProps } from '../../hooks';
+import { useScrollViewStyleProps } from './useScrollViewStyleProps';
+import { useScrollPosition } from './useScrollPosition';
+import { SCROLL_VIEW_DEFAULT_EDGE_INDICATOR, SCROLL_VIEW_DEFAULT_DIRECTION } from './constants';
+
+const ScrollView = (props: SpiritScrollViewProps) => {
+  const {
+    children,
+    direction = SCROLL_VIEW_DEFAULT_DIRECTION,
+    indicators = SCROLL_VIEW_DEFAULT_EDGE_INDICATOR,
```
Comment (literat) [inline packages/web-react/src/components/ScrollView/ScrollView.tsx:13]: I prefer using the `edge` prefix. I was asking by myself, what kind of indicators are they or what they indicate.

```suggestion
    edgeIndicators = SCROLL_VIEW_DEFAULT_EDGE_INDICATOR,
```

File: packages/web-react/src/components/ScrollView/__tests__/useScrollViewStyleProps.test.ts (hunk 14 above)
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/__tests__/useScrollViewStyleProps.test.ts:22]: done

File: packages/web-react/src/components/ScrollView/ScrollView.tsx (hunk 15 above)
Comment (literat) [reply packages/web-react/src/components/ScrollView/ScrollView.tsx:13]: But this is maybe also for @adamkudrna and @crishpeen because I did not notice it before that this is based on class names in the ScrollView component. I have noticed that in many places where there is a usage of `indicators` vs `EDGE_INDICATORS`. And it makes me no sense when we assign edge indicator to only indicator. Because there is an information loss. What do you think?

**Review by literat — APPROVED** [review]
Just thinking about naming consistency, but nothing blocking.

File: packages/web-react/src/components/ScrollView/ScrollView.tsx (hunk 15 above)
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/ScrollView.tsx:13]: I had this exactly set there, but when I looked at twig, I changed it to indicators

File: packages/web-react/src/types/scrollView.ts (hunk 2 above)
Comment (tomassychra) [reply packages/web-react/src/types/scrollView.ts]: sorry, done now

File: packages/web-react/src/components/ScrollView/ScrollView.tsx (hunk 15 above)
Comment (tomassychra) [reply packages/web-react/src/components/ScrollView/ScrollView.tsx]: done

Comment (literat) [reply packages/web-react/src/components/ScrollView/ScrollView.tsx]: I do not blame you for anything. The problem I see is that in Twig there is no usage of `edge`. I like the name "edge indicator" and I would like to prefer it over just "indicator". The problem I see here is that assign one to another. E.g. I would rather use `indicators = SCROLL_VIEW_DEFAULT_INDICATOR` or `edgeIndicators = SCROLL_VIEW_DEFAULT_EDGE_INDICATOR`. But not both. But I repeat, this is only nitpicking and thinking about naming consistency.
